src/contrib/mlir/core/pass/ng_dialect_fused_ops.cpp

```diff
@@ -0,0 +1,144 @@
+#include <mlir/ExecutionEngine/ExecutionEngine.h>
+#include <mlir/ExecutionEngine/OptUtils.h>
+#include <mlir/IR/AffineExpr.h>
+#include <mlir/IR/IntegerSet.h>
+#include <mlir/IR/MLIRContext.h>
+#include <mlir/IR/StandardTypes.h>
+#include <mlir/Pass/PassManager.h>
+#include <mlir/Target/LLVMIR.h>
+#include <mlir/Transforms/DialectConversion.h>
+#include <mlir/Transforms/Passes.h>
+
+#include <iostream>
+#define DEBUG_TYPE "mlir-compiler"
+#define PASS_NAME "ngraph_dialect_fusion"
+
+using llvm::SmallVector;
+using llvm::StringRef;
+using llvm::ArrayRef;
+
+using namespace ngraph;
+using namespace mlir;
+using namespace mlir::edsc;
+using namespace mlir::edsc::op;
+
+namespace mlir
+{
+    static Value createSgemmOp(
+        PatternRewriter& rewriter, Operation* old_op, Value input1, Value input2, Value input3)
+    {
+        auto castedOp0 = dyn_cast_or_null<NGAddOp>(old_op);
+        (void)castedOp0;
+        SmallVector<Value, 4> values;
+        (void)values;
+        SmallVector<NamedAttribute, 4> attrs;
+        (void)attrs;
+        values.push_back(input1);
+        values.push_back(input2);
+        values.push_back(input3);
+        attrs.emplace_back(
+            rewriter.getIdentifier("alpha"),
+            rewriter.getFloatAttr(mlir::Builder(rewriter.getContext()).getF32Type(), 1.0));
+        attrs.emplace_back(
+            rewriter.getIdentifier("beta"),
+            rewriter.getFloatAttr(mlir::Builder(rewriter.getContext()).getF32Type(), 1.0));
+        attrs.emplace_back(rewriter.getIdentifier("transA"), rewriter.getBoolAttr(false));
+        attrs.emplace_back(rewriter.getIdentifier("transB"), rewriter.getBoolAttr(false));
+        SmallVector<Type, 4> types;
+        (void)types;
+        for (auto v : castedOp0.getODSResults(0))
+        {
+            types.push_back(v.getType());
+        }
+        return rewriter.create<NGGemmOp>(castedOp0.getLoc(), types, values, attrs);
+    }
+
+#include "fused_ops_pattern.h.inc"
+}
+namespace
+{
+    class NgDialectFusedOpsPass : public mlir::ModulePass<NgDialectFusedOpsPass>
+    {
+    public:
+        NgDialectFusedOpsPass(mlir::ModuleOp& module, mlir::MLIRContext* context)
+            : m_module(module)
+            , m_context(context)
+        {
+        }
+
+        NgDialectFusedOpsPass(const NgDialectFusedOpsPass& obj);
+
+    private:
+        void runOnModule() override;
+        mlir::ModuleOp m_module;
+        mlir::MLIRContext* m_context;
+    };
+}
+
+NgDialectFusedOpsPass::NgDialectFusedOpsPass(const NgDialectFusedOpsPass& obj)
+    : m_module(obj.m_module)
+    , m_context(obj.m_context)
+{
+}
+
+void NgDialectFusedOpsPass::runOnModule()
+{
+    OwningRewritePatternList patterns;
+    mlir::populateWithGenerated(m_context, &patterns);
+
+    // Gather functions to be processed. Note that new functions will be added to module as part
+    // of the function signature conversion so we have to collect the original ones before hand.
```

Comment:
what signature conversion?

Reply:
it was referring to the functions, which will be modified after applying the pattern

Reply:
This fusion would not add new functions to the module. It only modifies the content of the original functions. Keeping only the first sentence is enough.